scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -85,21 +101,110 @@ object BeamOp extends Serializable {
     }
   }
 
-  final case class TransformBeamOp[A, B](source: BeamOp[A], f: PTransform[PCollection[A], PCollection[B]], kryoCoder: KryoCoder) extends BeamOp[B] {
+  final case class TransformBeamOp[A, B](
+    source: BeamOp[A],
+    f: PTransform[PCollection[A], PCollection[B]],
+    kryoCoder: KryoCoder
+  ) extends BeamOp[B] {
     def run(pipeline: Pipeline): PCollection[B] = {
       source.run(pipeline).asInstanceOf[PCollection[A]].apply(f).setCoder(kryoCoder)
     }
   }
 
+  final case class HashJoinOp[K, V, U, W](
+    left: BeamOp[(K, V)],
+    right: BeamOp[(K, U)], joiner: (K, V, Iterable[U]) => Iterator[W]
+  )(implicit kryoCoder: KryoCoder) extends BeamOp[(K, W)] {
+    override def run(pipeline: Pipeline): PCollection[_ <: (K, W)] = {
+      val leftPCollection = left.run(pipeline)
+      val rightPCollectionView = right.run(pipeline).asInstanceOf[PCollection[(K, U)]]
+        .apply(TupleToKV[K, U](kryoCoder, kryoCoder))
+        .apply(GroupByKey.create[K, U]()).setCoder(KvCoder.of(kryoCoder, kryoCoder))
+        .apply(View.asMap[K, java.lang.Iterable[U]]())
+
+      leftPCollection
+        .apply(
+          ParDo
+            .of(HashJoinFn[K, V, U, W](joiner, rightPCollectionView))
+            .withSideInputs(rightPCollectionView)
+        )
+        .setCoder(kryoCoder)
+    }
+  }
+
+  final case class CoGroupedOp[K, V](
+    cg: CoGrouped[K, V],
+    inputOps: Seq[BeamOp[(K, Any)]]
+  )(implicit kryoCoder: KryoCoder) extends BeamOp[(K, V)] {
```

Comment:
cg has `Ordering[K]`. We should use that to potentially get an `Coder[K]` 

https://github.com/twitter/scalding/blob/e445ae8d931845ddccec0795f2a0cb61c57d31d7/scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/KryoCoder.scala#L37